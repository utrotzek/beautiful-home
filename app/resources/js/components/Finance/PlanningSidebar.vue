<template>
    <div class="planningSidebar">
        <TemplateSelector
            :show="showTemplateSelector"
            @close="toggleTemplateSelector"
            @templateSelected="importTemplate"
        />
        <div class="row">
            <div class="col-8">
                <h2 class="d-none d-md-block">
                    {{ headline }}
                </h2>
            </div>
            <div class="col-12 col-md-4 text-left text-md-right">
                <div
                    id="PlanningButtons"
                    class="mb-3"
                >
                    <button
                        v-if="allowTemplateImport"
                        class="btn btn-outline-dark"
                        title="Einträge aus Vorlage übernehmen"
                        @click="toggleTemplateSelector"
                    >
                        <i class="fa fa-share-square"></i>
                        <span class="d-inline-block d-md-none">
                            Von Vorlage
                        </span>
                    </button>

                    <button
                        class="btn btn-outline-dark"
                        title="Neuen Planungs-Eintrag erstellen"
                        @click="createNewPlanning"
                    >
                        <i class="fa fa-plus-circle"></i>
                        <span class="d-inline-block d-md-none">
                            Neuer Eintrag
                        </span>
                    </button>
                </div>
            </div>
        </div>

        <div
            id="planningSearch"
            class="mb-3"
        >
            <Search @searched="planningSearched" />
        </div>

        <div
            v-for="(item) in filteredPlanning"
            :key="item.id"
        >
            <PlanningElement
                v-if="item.display && (item.totalAmount !== 0 || item.isNew)"
                :planning-item="item"
                :edit-mode="item.editMode"
                :year="year"
                :month="month"
                :cost-center-data="costCenterData"
                :allow-connection="allowConnection"
                @connect="connectPlanning(item.id, true)"
                @delete="deletePlanning(item.id)"
                @close="connectPlanning(item.id, false)"
                @save="savePlanning"
                @createCostCenter="createCostCenter"
            />
        </div>
        <div v-if="planningData.length !== 0 && filteredPlanning.length === 0">
            <div
                class="alert alert-light"
                role="alert"
            >
                Für den eingegebenen Suchbegriff wurden keine Einträge gefunden.
            </div>
        </div>
        <div
            v-if="noActivePlannings"
            class="alert alert-info"
            role="alert"
        >
            Keine Planungseinträge vorhanden
        </div>
    </div>
</template>

<script>
import _ from "lodash";
import moment from "moment";
import PlanningElement from "./PlanningElement";
import Search from "../tools/Search";
import TemplateSelector from "./Modals/TemplateSelector";

export default {
    name: "PlanningSidebar",
    components: {
        PlanningElement,
        Search,
        TemplateSelector
    },
    props: {
        headline: {
            type: String,
            required: false,
            default: "Planung"
        },
        year: {
            type: Number,
            required: true,
            default: null
        },
        month: {
            type: Number,
            required: true,
            default: null
        },
        planningData: {
            type: Array,
            required: true,
            default: null
        },
        costCenterData: {
            type: Array,
            required: true,
            default: null
        },
        allowConnection: {
            type: Boolean,
            required: false,
            default: true
        },
        allowTemplateImport: {
            type: Boolean,
            required: false,
            default: true
        }
    },
    data () {
        return {
            planningQuery: "",
            showTemplateSelector: false
        };
    },
    computed: {
        noActivePlannings() {
            let i = 0;
            for (i=0; i < this.planningData.length; i++){
                let planninigElement = this.planningData[i];
                if (planninigElement.totalAmount !== 0 && !planninigElement.isNew){
                    return false;
                }
            }
            return true;
        },
        orderedPlanning: function() {
            return _.orderBy(this.planningData, ["isNew", "date"], ["desc", "asc"]);
        },
        filteredPlanning() {
            //filters the planning list by title, description and date
            return this.orderedPlanning.filter(planning => {
                let costCenter = planning.costCenter;
                let display = (costCenter.title.toLowerCase().search(this.planningQuery.toLowerCase()) > -1);
                if (!display && planning.description){
                    display = (planning.description.toLowerCase().search(this.planningQuery.toLowerCase()) > -1);
                }
                if (!display) {
                    let formattedDate = moment(planning.date).format("DD.MM.YYYY");
                    display = (formattedDate.search(this.planningQuery.toLowerCase()) > -1);
                }
                return display;
            });
        },
    },
    methods: {
        savePlanning(planningItemToUpdate){
            this.$emit("save", planningItemToUpdate);
        },
        planningSearched(query) {
            this.planningQuery = query;
        },
        createNewPlanning() {
            this.$emit("createNewPlanning");
        },
        deletePlanning(id) {
            this.$emit("deletePlanning", id);
        },
        connectPlanning(id, enabled) {
            this.$emit("connectPlanning", id, enabled);
        },
        createCostCenter(newCostCenter) {
            this.$emit("createCostCenter", newCostCenter);
        },
        toggleTemplateSelector() {
            this.showTemplateSelector = !this.showTemplateSelector;
        },
        importTemplate(template) {
            this.$emit("importTemplate", template);
        }
    }
};
</script>

<style scoped>

</style>
