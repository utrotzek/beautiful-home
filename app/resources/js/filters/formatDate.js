import moment from "moment";

//TODO: set format in moment
export default function formatDate (value) {
    if (value) {
        if (typeof value === "object"){
            return moment(value.date).format("DD.MM.YYYY");
        }else{
            return moment(String(value)).format("DD.MM.YYYY");
        }
    }
}
