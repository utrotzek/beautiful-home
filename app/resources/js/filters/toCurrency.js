export default function toCurrency (value) {
    if (typeof value !== "number") {
        return value;
    }
    var formatter = new Intl.NumberFormat("de-DE", {
        style: "currency",
        currency: "EUR",
        minimumFractionDigits: 2
    });
    return formatter.format(value);
}
