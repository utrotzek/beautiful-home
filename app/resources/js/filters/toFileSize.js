export default function toFileSize(value){
    if (isNaN(value)){
        return "undefined size";
    }else{
        const units = [ "byte", "kB", "MB", "GB", "TB"];
        let newValue = value;
        let counter = 0;

        if (newValue >= 1000){
            while (newValue >= 1000){
                newValue = newValue / 1024;
                counter++;
            }
        }

        var formatter = new Intl.NumberFormat("de-DE", {
            minimumFractionDigits: 1,
            maximumFractionDigits: 1
        });
        return formatter.format(newValue) + " " + units[counter];
    }
}
