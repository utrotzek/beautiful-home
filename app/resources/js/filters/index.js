import toCurrency from "./toCurrency";
import formatDate from "./formatDate";
import toFileSize from "./toFileSize";

export default {
    install(Vue) {
        Vue.filter("toCurrency", toCurrency);
        Vue.filter("formatDate", formatDate);
        Vue.filter("toFileSize", toFileSize);
    }
};
