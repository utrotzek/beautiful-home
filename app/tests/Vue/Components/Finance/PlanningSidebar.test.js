import "../Mocks/MatchMedia.mock";
import _ from "lodash";
import {shallowMount} from "@vue/test-utils";
import PlanningSidebar from "../../../../resources/js/components/Finance/PlanningSidebar";
import Vue from "vue";
import filters from "../../../../resources/js/filters";
import PlanningDataFixtures from "./Fixtures/PlanningFixtures";
import CostCenterFixtures from "./Fixtures/CostCenterFixtures";
import TemplateFixtures from "./Fixtures/TemplateFixtures";

Vue.use(filters);

describe("Planning sidebar", () => {
    let wrapper = null;
    let planningData = null;
    let costCenterData = null;

    beforeEach(() => {
        planningData = PlanningDataFixtures;
        costCenterData = CostCenterFixtures;

        wrapper = shallowMount(PlanningSidebar,{
            propsData: {
                headline: "Fixture headline",
                year: 3000,
                month: 1,
                planningData: planningData,
                costCenterData: costCenterData,
                allowConnect: true,
                allowTemplateImport: false
            }
        });
    });

    it ("renders properly", () => {
        expect(wrapper.element).toMatchSnapshot();
    });

    it ("can toggle template selector", async () => {
        expect(wrapper.vm.showTemplateSelector).toBeFalsy();
        wrapper.vm.toggleTemplateSelector();
        expect(wrapper.vm.showTemplateSelector).toBeTruthy();
        wrapper.vm.toggleTemplateSelector();
        expect(wrapper.vm.showTemplateSelector).toBeFalsy();
    });

    it ("fires import event when importing template", async () => {
        const expectedTemplate = TemplateFixtures[0];

        wrapper.vm.importTemplate(expectedTemplate);
        expect(wrapper.emitted().importTemplate[0]).toEqual([expectedTemplate]);
    });

    it("fires create event when creating cost center", async() => {
        const expectedCostCenter = "My new cost center";
        wrapper.vm.createCostCenter(expectedCostCenter);
        expect(wrapper.emitted().createCostCenter[0]).toEqual([expectedCostCenter]);
    });

    it ("fires connect planning event when connecting planning", async () => {
        wrapper.vm.connectPlanning(999, true);
        expect(wrapper.emitted().connectPlanning[0]).toEqual([999, true]);
    });

    it ("fires delete planning event when deleting planning", async () => {
        wrapper.vm.deletePlanning(999);
        expect(wrapper.emitted().deletePlanning[0]).toEqual([999]);
    });

    it ("fires create planning event when deleting planning", async () => {
        wrapper.vm.createNewPlanning();
        expect(wrapper.emitted().createNewPlanning).toBeTruthy();
    });

    it ("fires save event when saving planning", async () => {
        const expectedSavedPlanning = {id: 1, title: "savedPlanning", totalAmount: -100};
        wrapper.vm.savePlanning(expectedSavedPlanning);
        expect(wrapper.emitted().save[0]).toEqual([expectedSavedPlanning]);
    });

    it ("can filter plannings by title", async () => {
        const expectedPlannings = [
            planningData[0],
            planningData[1]
        ];

        wrapper.vm.planningSearched("Einkauf");
        const filteredPlannings = wrapper.vm.filteredPlanning;
        expect(filteredPlannings).toEqual(expectedPlannings);
    });

    it ("can filter plannings by description", async () => {
        const expectedPlannings = [
            planningData[2],
            planningData[3]
        ];

        wrapper.vm.planningSearched("person 1");
        const filteredPlannings = wrapper.vm.filteredPlanning;
        expect(filteredPlannings).toEqual(expectedPlannings);
    });

    it ("displays a text if no planning entries are active", async () => {
        const fixtureData = [
            {
                id: 1,
                costCenter: {
                    id: 1,
                    title: "Einkaufen"
                },
                description: "Eintrag 1",
                totalAmount: 0,
                isNew: false
            }
        ];
        wrapper.setProps({planningData: fixtureData});
        await Vue.nextTick();

        expect(wrapper.text()).toContain("Keine Planungseinträge vorhanden");
    });

    it ("displays a text if query doesn't match any entries", async () => {
        wrapper.vm.planningSearched("This query string does not exist");
        await Vue.nextTick();
        expect(wrapper.text()).toContain("Für den eingegebenen Suchbegriff wurden keine Einträge gefunden");
    });

    it ("can determine if click of planning item should be enabled", async () =>{
        wrapper.vm.connectPlanningMode = false;
        wrapper.vm.connectPlanningId = 100;
    });
});
