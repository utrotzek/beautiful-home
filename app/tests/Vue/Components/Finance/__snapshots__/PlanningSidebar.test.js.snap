// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Planning sidebar renders properly 1`] = `
<div
  allowconnect="true"
  class="planningSidebar"
>
  <templateselector-stub />
   
  <div
    class="row"
  >
    <div
      class="col-8"
    >
      <h2
        class="d-none d-md-block"
      >
        
                Fixture headline
            
      </h2>
    </div>
     
    <div
      class="col-12 col-md-4 text-left text-md-right"
    >
      <div
        class="mb-3"
        id="PlanningButtons"
      >
        <!---->
         
        <button
          class="btn btn-outline-dark"
          title="Neuen Planungs-Eintrag erstellen"
        >
          <i
            class="fa fa-plus-circle"
          />
           
          <span
            class="d-inline-block d-md-none"
          >
            
                        Neuer Eintrag
                    
          </span>
        </button>
      </div>
    </div>
  </div>
   
  <div
    class="mb-3"
    id="planningSearch"
  >
    <search-stub />
  </div>
   
  <div>
    <!---->
  </div>
  <div>
    <!---->
  </div>
  <div>
    <!---->
  </div>
  <div>
    <!---->
  </div>
   
  <!---->
   
  <!---->
</div>
`;
